README.rst
=============
python-recsys
=============

A python library for implementing a recommender system.

- Now supports incrementally adding new users or items instead of building the model from scratch for these new users or items via the folding-in technique which was mentioned in Sarwar et al.'s `paper`_ (Titled: Incremental Singular Value Decomposition Algorithms for Highly Scalable Recommender Systems), this latest commit is simply an implementation to it for python-recsys.

.. _`paper`: http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.3.7894&rep=rep1&type=pdf

- A `Demonstration video is available`_  for this latest commit in form of a demo site built using the MEAN stack which uses the updated python-recsys as backend for the recommender which folds-in the website's user in to the SVD model and gets recommendations instantaneously instead of building the model from scratch.

.. _`Demonstration video is available`:  https://youtu.be/tIvQxBfa2d4

-There is also an accompanying `bachelor thesis paper`_ (For those interested) which outlines the background, architecture and discusses the "Folding-in" approach.

.. _`bachelor thesis paper`: https://drive.google.com/file/d/0BylQe2cRVWE_RmZoUTJYSGZNaXM/view


Installation
============

Dependencies
~~~~~~~~~~~~

**python-recsys** is build on top of `Divisi2`_, with csc-pysparse (Divisi2 also requires `NumPy`_, and uses Networkx).

.. _`Divisi2`: http://csc.media.mit.edu/docs/divisi2/install.html
.. _`NumPy`: http://numpy.scipy.org

**python-recsys** also requires `SciPy`_.

.. _`SciPy`: http://numpy.scipy.org

To install the dependencies do something like this (Ubuntu):

::

    sudo apt-get install python-scipy python-numpy
    sudo apt-get install python-pip
    sudo pip install csc-pysparse networkx divisi2

    # If you don't have pip installed then do:
    # sudo easy_install csc-pysparse
    # sudo easy_install networkx
    # sudo easy_install divisi2

Download
~~~~~~~~

Download **python-recsys**  from `github`_.

.. _`github`: http://github.com/ocelma/python-recsys

Install
~~~~~~~

::

    tar xvfz python-recsys.tar.gz
    cd python-recsys
    sudo python setup.py install

Example
~~~~~~~

1. Load Movielens dataset:

::

    from recsys.algorithm.factorize import SVD
    svd = SVD()
    svd.load_data(filename='./data/movielens/ratings.dat',
                sep='::',
                format={'col':0, 'row':1, 'value':2, 'ids': int})

2. Compute Singular Value Decomposition (SVD), M=U Sigma V^t:

::

    k = 100
    svd.compute(k=k,
                min_values=10,
                pre_normalize=None,
                mean_center=True,
                post_normalize=True,
                savefile='/tmp/movielens')

3. Get similarity between two movies:

::

    ITEMID1 = 1    # Toy Story (1995)
    ITEMID2 = 2355 # A bug's life (1998)

    svd.similarity(ITEMID1, ITEMID2)
    # 0.67706936677315799

4. Get movies similar to *Toy Story*:

::

    svd.similar(ITEMID1)

    # Returns: <ITEMID, Cosine Similarity Value>
    [(1,    0.99999999999999978), # Toy Story
     (3114, 0.87060391051018071), # Toy Story 2
     (2355, 0.67706936677315799), # A bug's life
     (588,  0.5807351496754426),  # Aladdin
     (595,  0.46031829709743477), # Beauty and the Beast
     (1907, 0.44589398718134365), # Mulan
     (364,  0.42908159895574161), # The Lion King
     (2081, 0.42566581277820803), # The Little Mermaid
     (3396, 0.42474056361935913), # The Muppet Movie
     (2761, 0.40439361857585354)] # The Iron Giant

5. Predict the rating a user (USERID) would give to a movie (ITEMID):

::

    MIN_RATING = 0.0
    MAX_RATING = 5.0
    ITEMID = 1
    USERID = 1

    svd.predict(ITEMID, USERID, MIN_RATING, MAX_RATING)
    # Predicted value 5.0

    svd.get_matrix().value(ITEMID, USERID)
    # Real value 5.0

6. Recommend (non-rated) movies to a user:

::

    svd.recommend(USERID, is_row=False) #cols are users and rows are items, thus we set is_row=False

    # Returns: <ITEMID, Predicted Rating>
    [(2905, 5.2133848204673416), # Shaggy D.A., The
     (318,  5.2052108435956033), # Shawshank Redemption, The
     (2019, 5.1037438278755474), # Seven Samurai (The Magnificent Seven)
     (1178, 5.0962756861447023), # Paths of Glory (1957)
     (904,  5.0771405690055724), # Rear Window (1954)
     (1250, 5.0744156653222436), # Bridge on the River Kwai, The
     (858,  5.0650911066862907), # Godfather, The
     (922,  5.0605327279819408), # Sunset Blvd.
     (1198, 5.0554543765500419), # Raiders of the Lost Ark
     (1148, 5.0548789542105332)] # Wrong Trousers, The

7. Which users should *see* Toy Story? (e.g. which users -that have not rated Toy
   Story- would give it a high rating?)

::

    svd.recommend(ITEMID)

    # Returns: <USERID, Predicted Rating>
    [(283,  5.716264440514446),
     (3604, 5.6471765418323141),
     (5056, 5.6218800339214496),
     (446,  5.5707524860615738),
     (3902, 5.5494529168484652),
     (4634, 5.51643364021289),
     (3324, 5.5138903299082802),
     (4801, 5.4947999354188548),
     (1131, 5.4941438045650068),
     (2339, 5.4916048051511659)]


Example for incremental update
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1. Load Movielens dataset and prepare for training and testing:

::

    import recsys.algorithm
    recsys.algorithm.VERBOSE = True

    from recsys.algorithm.factorize import SVD
    from recsys.datamodel.data import Data

    filename = “(your movielens file path here)”

    #In movielens dataset, the user is at 0 so I made them the row (could keep it as above {'col':0, 'row':1, 'value':2, 'ids': int} but I changed order to emphasis a parameter in an upcoming function)
    format = {'col':1, 'row':0, 'value':2, 'ids': int}

    data = Data()
    data.load(filename, sep='::', format=format)
    #splits the dataset according to row or column (based on is_row=true or false) which causes there to be no overlap (of users for example) between train and foldin dataset
    train, test, foldin = data.split_train_test_foldin(base=60,percentage_base_user=80,shuffle_data=True,is_row=True) #since users are in the row so is_row=true

    # Returns: a tuple <Data, Data, Data> for train, test, foldin
    # Prints: (If VERBOSE=True)
    total number of tuples: 1000209
    percentage of data for training: 48.0 % with 479594 tuples
    percentage of data for testing: 20.0 % with 200016 tuples # 100-percentage_base_user per user (percentage of tuples which means the ratings since a user has many tuples(ratings))
    percentage of data for foldin: 32.0 % with 320599 tuples
    _____________
    percentage of users for foldin: 40.0 % with 2416 users # 100-base= foldin (percentage of users)
    percentage of users for training: 60.0 % with 3624 users #base for training (percentage of users)

2. Compute Singular Value Decomposition (SVD), M=U Sigma V^t:

::

    svd = SVD()
    svd.set_data(train)
    svd.compute(k=100,
            	min_values=1,
            	pre_normalize=None,
            	mean_center=False,
            	post_normalize=True)

    # Prints:
    Creating matrix (479594 tuples)
    Matrix density is: 3.7007%
    Updating matrix: squish to at least 1 values
    Computing svd k=14, min_values=1, pre_normalize=None, mean_center=False, post_normalize=False

3. "Foldin" those new users or items (update model instead of updating from scratch)

::

    svd.load_updateDataBatch_foldin(data=foldin,is_row=True)

    # Prints: (If VERBOSE=True)
    before updating, M= (3624, 3576)
    done updating, M= (6040, 3576) # Folds in all the new users (not previously in model)

4. Recommend (non-rated) movies to a NEW user
::

    user_id=foldin[0][1] #returns userID which is in foldin dataset BUT not in train dataset
    svd.recommend(user_id,is_row=True,only_unknowns=True) #The userID is in row and gets only the unrated (unknowns)

    # Returns: <ITEMID, Predicted Rating>
    [(1307, 3.6290483094468913),
    (1394, 3.5741565545425957),
    (1259, 3.5303836262378048),
    (1968, 3.4565426585553927),
    (2791, 3.3470277643217203),
    (1079, 3.268283171487782),
    (1198, 3.2381080336246675),
    (593, 3.204915630088236),
    (1270, 3.1859618303393233),
    (2918, 3.1548530640630252)]

5. Recommend (non-rated) movies to a NEW user and validate not in base model (prior to folding-in)
::

    # BEFORE running points 3 and 4 (prior to calling svd.load_updateDataBatch_foldin)

    user_id=foldin[0][1] #returns userID which is in foldin dataset BUT not in train dataset

    # Try block to validate that the userID is new and not in the base model
    try:
        print "Getting recommendation for user_id which was not in original model training set"
        print "recommendations:",svd.recommend(user_id)
    except Exception:
        print "New user not in base model so in except block and will foldin the foldin dataset (update the model NOT calculate from scratch)"
        svd.load_updateDataBatch_foldin(data=foldin,format=format,is_row=True,truncate=True,post_normalize=True)
        print "recommendations:",svd.recommend(user_id,is_row=True,only_unknowns=True) #The userID is in row and get us only the unrated (unknowns)


    # Prints:
    Getting recommendation for user_id which was not in original model training set
    recommendations: New user not in base model so in except block and will foldin the foldin dataset (update the model NOT calculate from scratch)
    before updating, M= (3624, 3576)
    done updating, M= (6040, 3576)
    recommendations: [(1307, 3.6290483094468913), (1394, 3.5741565545425957), (1259, 3.5303836262378048), (1968, 3.4565426585553927), (2791, 3.3470277643217203), (1079, 3.268283171487782), (1198, 3.2381080336246675), (593, 3.204915630088236), (1270, 3.1859618303393233), (2918, 3.1548530640630252)]


6. Load previous SVD model and foldin NEW users from file then instantly get recommendations
::

    format = {'col':1, 'row':0, 'value':2, 'ids': int}

    svd = SVD()
    #load base svd model
    svd.load_model('SVDModel')

    # load new users by their movie rating data file and use it to fold-in the users into the model (loads data and folds in)
    svd.load_updateDataBatch_foldin(filename = 'newUsers.dat', sep='::', format=formate, is_row=True)

    # gets recommendedations
    print "recommendations:", svd.recommend(new_userID,is_row=True,only_unknowns=True)


- All the normal functionalities of python-recsys are compatible with the incremental update commit. The incremental update can even work if you load the model then foldin a new user or users or even items.

- Please note that preexisting users can't be folded-in only new users which aren't already in the svd model.

Documentation
~~~~~~~~~~~~~

Documentation and examples available `here`_.

.. _`here`: http://ocelma.net/software/python-recsys/build/html

To create the HTML documentation files from doc/source do:

::

    cd doc
    make html

HTML files are created here:

::

    doc/build/html/index.html


CHANGELOG
See: https://github.com/ocelma/python-recsys/commits/master

2011-05-09
    * Added MAP (Mean Average Precision) metric and (nose)tests

2011-02-01
    * Added external SVDLIBC executable in recsys.utils
    * Load larger matrices using SVDLIBC instead of divisi2
    * SVDLIBC examples

2011-01-23
    * Improved(?) K-Means clustering and added a test

2011-01-12
    * added recsys.utils: compute SVDLIBC [http://tedlab.mit.edu/~dr/SVDLIBC/], and import U,S,V to recsys.algorithm.factorize.SVD
    * added new test_utils.py to check SVDLIBC to recsys import

2011-11-14
    * add_tuple raises more exceptions, if tuple format not correct

2011-11-08
    * better explanation of the format parameter in load_data
    * save shifted matrix only if mean_center is True

2011-10-14
    * loads and saves zip bigger than 4Gb (allow64=True)
    * loading data from file (load_data method) more robust now

2011-10-08
    * Added the whole project at github

2017-06-08
    * Added updating the SVD model incrementally by folding-in
    * Added a capability to split the dataset for train,test,foldin to facilitate testing the Fold-in implementation
    * Added update to github


recsys/algorithm/matrix.py
try:
    from divisi2.sparse import SparseMatrix as divisiSparseMatrix
    from divisi2 import reconstruct_similarity
except:
    from csc.divisi2.sparse import SparseMatrix as divisiSparseMatrix
    from csc.divisi2 import reconstruct_similarity

from operator import itemgetter

class Matrix(object):
    def __init__(self):
        self._matrix = None

    def __repr__(self):
        return str(self._matrix)

    def create(self, data):
        raise NotImplementedError("cannot instantiate Abstract Base Class")

    def density(self, percent=True):
        if not self._matrix or not self._matrix.entries():
            return None
        density = self._matrix.density()
        if percent:
            density *= 100
        return round(density, 4)

    def empty(self):
        raise NotImplementedError("cannot instantiate Abstract Base Class")

    def get(self):
        return self._matrix

    def set(self, matrix):
        self._matrix = matrix

    def get_row(self, i):
        if self.empty() or not self._matrix.col_labels:
            raise ValueError('Matrix is empty (or has no columns!)')
        return self._matrix.row_named(i)

    def get_col(self, j):
        if self.empty() or not self._matrix.row_labels:
            raise ValueError('Matrix is empty (or has no rows!)')
        return self._matrix.col_named(j)

    def value(self, i, j):
        if self.empty():
            raise ValueError('Matrix is empty!')
        return self._matrix.entry_named(i, j)

    def get_value(self, i, j):
        if self.empty():
            raise ValueError('Matrix is empty!')
        return self.value(i, j)

    def set_value(self, i, j, value):
        if self.empty():
            raise ValueError('Matrix is empty!')
        self._matrix.set_entry_named(i, j, value)

    def get_row_len(self):
        if self.empty() or not self._matrix.col_labels:
            raise ValueError('Matrix is empty (or has no columns!)')
        return len(self._matrix.col_labels)

    def get_col_len(self):
        if self.empty() or not self._matrix.row_labels:
            raise ValueError('Matrix is empty (or has no rows!)')
        return len(self._matrix.row_labels)


class SparseMatrix(Matrix):
    def __init__(self):
        super(SparseMatrix, self).__init__()
        self._values=None
        self._rows=None
        self._cols=None
        self._additional_elements=[] #if no additional then len will equal 0


    def get_rows(self): #can use to get rated items and remove from recommendation
        return self._rows

    def get_cols(self): #can use to get rated items and remove from recommendation
        return self._cols

    def get_additional_elements(self):  # can use to get additional items to either fold or truncate
        return self._additional_elements

#row_labels specifies the row labels the complete matrix should have incase the inputted file doesn't include all indicies and it was saved in previous matrix (for update)
#same explination for col_labels but for columns
#matrix should have, in case it is larger than the largest index.
    def create(self, data,row_labels=None, col_labels=None, foldin=False,truncate=False):
         #is_row is what I'm originally folding in
        self._values = map(itemgetter(0), data)
        self._rows = map(itemgetter(1), data)
        self._cols = map(itemgetter(2), data)

        if foldin: #new to make sure not folding in user and item at same time
            #idea: create matrix normally but keep track of the columns (items) or rows to be folded in before doing update
            if col_labels: #if col_labels defined then I'm folding in a row
                self._additional_elements = [x for x in self._cols if x not in col_labels]
            else: #else I am folding in a column
                self._additional_elements = [x for x in self._rows if x not in row_labels]
            if truncate:
                for item in self._additional_elements:
                    if col_labels:
                        index_remove = self._cols.index(item)
                    else:
                        index_remove = self._rows.index(item)
                    del self._values[index_remove]
                    del self._rows[index_remove]
                    del self._cols[index_remove]


        self._matrix = divisiSparseMatrix.from_named_lists(self._values, self._rows, self._cols,row_labels, col_labels)




    def update(self, matrix,is_batch=False): #isbatch is for creating the final sparse matrix ,since you will want to collect all then construct final matrix at end
#To update the stored data matrix with the new values and create a new divisi spare matrix with it to retain the zeroes
        self._values.extend(matrix._values)
        self._rows.extend(matrix._rows)
        self._cols.extend(matrix._cols)

        if not is_batch:
            self._matrix = divisiSparseMatrix.from_named_lists(self._values, self._rows, self._cols)

    def squish(self,squishFactor): #remove additional empty fields created by divisiSparseMatrix
        self._matrix=self._matrix.squish(squishFactor)

    def index_sparseMatrix(self): #create the divisi2 sparse matrix from already existing values
        self._matrix = divisiSparseMatrix.from_named_lists(self._values, self._rows, self._cols)

    def empty(self):
        return not self._matrix or not self._matrix.values()

class SimilarityMatrix(Matrix):
    def __init__(self):
        super(SimilarityMatrix, self).__init__()

    def create(self, U, S, post_normalize=False):
        self._matrix = reconstruct_similarity(U, S, post_normalize=post_normalize)

    def empty(self):
        nrows, ncols = (0, 0)
        if self._matrix:
            nrows, ncols = self._matrix.shape
        return not self._matrix or not (nrows and ncols)


recsys/datamodel/data.py
import sys
import codecs
import pickle
#from random import shuffle
from exceptions import ValueError
from numpy.random import shuffle

from recsys.algorithm import VERBOSE

class Data:
    """
    Handles the relationshops among users and items
    """
    def __init__(self):
        #"""
        #:param data: a list of tuples
        #:type data: list
        #"""
        self._data = list([])
        self._tupleDict = {}

    def __repr__(self):
        s = '%d rows.' % len(self.get())
        if len(self.get()):
            s += '\nE.g: %s' % str(self.get()[0])
        return s

    def __len__(self):
        return len(self.get())

    def __getitem__(self, i):
        if i < len(self._data):
            return self._data[i]
        return None

    def __iter__(self):
        return iter(self.get())

    def set(self, data, extend=False):
        """
        Sets data to the dataset

        :param data: a list of tuples
        :type data: list
        """
        if extend:
            self._data.extend(data)
        else:
            self._data = data

    def get(self):
        """
        :returns: a list of tuples
        """
        return self._data
    def get_tuple_dict(self):
        """
        :returns: a dictionary of users or items and corresponding ratings
        """
        if not self._tupleDict:
            raise ValueError('Tuple dictionary hasn\'t been created yet, please run split_train_test_foldin first then try again')
        return self._tupleDict

    def add_tuple(self, tuple):
        """
        :param tuple: a tuple containing <rating, user, item> information (e.g.  <value, row, col>)
        """
        #E.g: tuple = (25, "ocelma", "u2") -> "ocelma has played u2 25 times"
        if not len(tuple) == 3:
            raise ValueError('Tuple format not correct (should be: <value, row_id, col_id>)')
        value, row_id, col_id = tuple
        if not value and value != 0:
            raise ValueError('Value is empty %s' % (tuple,))
        if isinstance(value, basestring):
            raise ValueError('Value %s is a string (must be an int or float) %s' % (value, tuple,))
        if row_id is None or row_id == '':
            raise ValueError('Row id is empty %s' % (tuple,))
        if col_id is None or col_id == '':
            raise ValueError('Col id is empty %s' % (tuple,))
        self._data.append(tuple)

    def split_train_test(self, percent=80, shuffle_data=True):
        """
        Splits the data in two disjunct datasets: train and test

        :param percent: % of training set to be used (test set size = 100-percent)
        :type percent: int
        :param shuffle_data: shuffle dataset?
        :type shuffle_data: Boolean

        :returns: a tuple <Data, Data>
        """
        if shuffle_data:
            shuffle(self._data)
        length = len(self._data)
        train_list = self._data[:int(round(length*percent/100.0))]
        test_list = self._data[-int(round(length*(100-percent)/100.0)):]
        train = Data()
        train.set(train_list)
        test = Data()
        test.set(test_list)

        return train, test

    def split_train_test_foldin(self,base=60,percentage_base_user=80, shuffle_data=True,is_row=True,force=True,data_report_path=None,id=None,ignore_rating_count=0):
        """
        Splits the data in three datasets: train, test, and foldin

        :param base: % of training set to be used (Foldin set size = 100-base) for base SVD model (not folded)
        :type base: int
        :param percentage_base_user: % of user ratings per user (or item ratings per item depending on which is row and column) to be used as base for training or foldin (testing will be percentage of ratings from 100-percentage_base_user per user or item )
        :type percentage_base_user: int

        :param shuffle_data: shuffle dataset?
        :type shuffle_data: Boolean

        :param is_row: are you trying to foldin a row or a column ? yes-> row , no-> column
        :type is_row: Boolean
        :param force: clear the values in data
        :type force: Boolean


        The following parameters are used for when generating a report of the dataset distribution:
        :param data_report_path: path to create report in
        :type data_report_path: String
        :param id: id number to be given to the report
        :type id: String
        :param ignore_rating_count: The threshold number of ratings to be removed from the data.
        :type ignore_rating_count: int

        :returns: a tuple <Data, Data, Data> for train, test, foldin
        """
        if force:
            self._construct_dictionary(is_row=is_row,force=True)
        elif len(self._tupleDict)==0:
            self._construct_dictionary(is_row=is_row)
        self._remove_ratings_count_from_dictionary(ignore_rating_count)
        dictKeys=self._tupleDict.keys() #users
        numberOfKeys= len(dictKeys) #number of users

        train_list =[]
        test_list=[]
        foldin_list=[]

        if shuffle_data:
            shuffle(dictKeys)
        train_list_keys=dictKeys[:int(round(numberOfKeys*base/100.0))]
        if base==100:
            foldin_list_keys=[]
        else:
            foldin_list_keys=dictKeys[-int(round(numberOfKeys*(100-base)/100.0)):]

        for key in train_list_keys:
            tupleList=self._tupleDict[key]
            lengthTupleList=len(tupleList)
            if shuffle_data:
                shuffle(tupleList)

            train_list.extend(tupleList[:int(round(lengthTupleList*percentage_base_user/100.0))])
            if int(round(lengthTupleList*(100-percentage_base_user)/100.0)) !=0: #if test=0 then can't take that percentage so skip taking it's tuple for test
                test_list.extend(tupleList[-int(round(lengthTupleList*(100-percentage_base_user)/100.0)):])

        for key in foldin_list_keys:
            tupleList=self._tupleDict[key]
            lengthTupleList=len(tupleList)
            if shuffle_data:
                shuffle(tupleList)

            foldin_list.extend(tupleList[:int(round(lengthTupleList*percentage_base_user/100.0))])
            if int(round(lengthTupleList*(100-percentage_base_user)/100.0)) !=0: #if test=0 then can't take that percentage so skip taking it's tuple for test
                test_list.extend(tupleList[-int(round(lengthTupleList*(100-percentage_base_user)/100.0)):])



        length = len(self._data)
        if VERBOSE:
            print "total number of tuples:",length
            print "percentage of data for training:",round((len(train_list)*1.0/length)*100),"%","with",len(train_list),"tuples"
            print "percentage of data for testing:",round((len(test_list)*1.0/length)*100),"%","with",len(test_list),"tuples"
            print "percentage of data for foldin:",round((len(foldin_list)*1.0/length)*100),"%","with",len(foldin_list),"tuples"
            print "_____________"
            print "percentage of users for foldin:",round((len(foldin_list_keys)*1.0/numberOfKeys*1.0)*100),"%","with",len(foldin_list_keys),"users"
            print "percentage of users for training:",round((len(train_list_keys)*1.0/numberOfKeys*1.0)*100),"%","with",len(train_list_keys),"users"

        if data_report_path:
            myFile = open(data_report_path+"/data_distribution_report.txt", 'a+')

            myFile.write("DataID:"+ str(id))
            myFile.write("total number of tuples:"+ str(length))
            myFile.write("\n")
            myFile.write( "percentage of data for training:"+ str(round((len(train_list) * 1.0 / length) * 100))+ "%"+ "with"+str(len(train_list))+"tuples")
            myFile.write("\n")
            myFile.write( "percentage of data for testing:"+ str(round((len(test_list) * 1.0 / length) * 100))+ "%"+ "with"+ str(len(test_list))+ "tuples")
            myFile.write("\n")
            myFile.write("percentage of data for foldin:"+ str(round((len(foldin_list) * 1.0 / length) * 100))+ "%"+ "with"+ str(len(foldin_list))+ "tuples")
            myFile.write("\n")
            myFile.write("_____________")
            myFile.write("\n")
            myFile.write("percentage of users for foldin:"+ str(round((len(foldin_list_keys) * 1.0 / numberOfKeys * 1.0) * 100))+ "%"+ "with"+ str(len(foldin_list_keys))+ "users")
            myFile.write("\n")
            myFile.write("percentage of users for training:"+ str(round((len(train_list_keys) * 1.0 / numberOfKeys * 1.0) * 100))+ "%"+"with"+ str(len(train_list_keys))+ "users")
            myFile.write("\n")
            myFile.write("________________________________________________________________")
            myFile.write("\n")

            myFile.close()


        train = Data()
        train.set(train_list)
        test = Data()
        test.set(test_list)
        foldin=Data()
        foldin.set(foldin_list)

        return train, test, foldin

    def _remove_ratings_count_from_dictionary(self,count_threshold_to_remove):
        '''
        :param count_threshold_to_remove: The threshold number of ratings to be removed from the data.
        :type count_threshold_to_remove: int
        :return: void, it changes the data itself in the class.
        '''
        if count_threshold_to_remove==0:
            return
        removed=0
        dictKeys=self._tupleDict.keys()
        for key in dictKeys:
            if len(self._tupleDict[key])<=count_threshold_to_remove:
                del self._tupleDict[key]
                removed+=1

        print "users removed less than or equal threshold count=",removed,"users"
        return

    def _construct_dictionary(self, is_row=True,force=True):
        '''

        :param data: Data()
        :param is_row: Boolean
        :return: constructs a dictionary with the row or col as the keys (depending on which is being added) with values as the tuples
        in self._batchDict
        '''
        # self._values = map(itemgetter(0), data)
        # self._rows = map(itemgetter(1), data)
        # self._cols = map(itemgetter(2), data)
        key_idx = 1  # key index default is the row
        if not is_row:
            key_idx = 2
        if force:    #construct new dictionary
            self._tupleDict={}
        # collecting the significant col or row tuples at one place to fold them in at once

        for item in self._data:  # data is a list of tuples so item is 1 tuple
            try:
                self._tupleDict[item[key_idx]].append(item)
            except KeyError:
                self._tupleDict[item[key_idx]] = []
                self._tupleDict[item[key_idx]].append(item)

        # batch loaded , now need to fold them in one by one
        if VERBOSE:
            print "Dictionary created successfully"

    def load(self, path, force=True, sep='\t', format=None, pickle=False):
        """
        Loads data from a file

        :param path: filename
        :type path: string
        :param force: Cleans already added data
        :type force: Boolean
        :param sep: Separator among the fields of the file content
        :type sep: string
        :param format: Format of the file content.
            Default format is 'value': 0 (first field), then 'row': 1, and 'col': 2.
            E.g: format={'row':0, 'col':1, 'value':2}. The row is in position 0,
            then there is the column value, and finally the rating.
            So, it resembles to a matrix in plain format
        :type format: dict()
        :param pickle: is input file in  pickle format?
        :type pickle: Boolean
        """
        if VERBOSE:
            sys.stdout.write('Loading %s\n' % path)
        if force:
            self._data = list([])
        if pickle:
            self._load_pickle(path)
        else:
            i = 0
            for line in codecs.open(path, 'r', 'ISO-8859-1'): #was utf8 changed it to 'ISO-8859-1'
                data = line.strip('\r\n').split(sep)
                value = None
                if not data:
                    raise TypeError('Data is empty or None!')
                if not format:
                    # Default value is 1
                    try:
                        value, row_id, col_id = data
                    except:
                        value = 1
                        row_id, col_id = data
                else:
                    try:
                        # Default value is 1
                        try:
                            value = data[format['value']]
                        except KeyError, ValueError:
                            value = 1
                        try:
                            row_id = data[format['row']]
                        except KeyError:
                            row_id = data[1]
                        try:
                            col_id = data[format['col']]
                        except KeyError:
                            col_id = data[2]
                        row_id = row_id.strip()
                        col_id = col_id.strip()
                        if format.has_key('ids') and (format['ids'] == int or format['ids'] == 'int'):
                            try:
                                row_id = int(row_id)
                            except:
                                print 'Error (ID is not int) while reading: %s' % data #Just ignore that line
                                continue
                            try:
                                col_id = int(col_id)
                            except:
                                print 'Error (ID is not int) while reading: %s' % data #Just ignore that line
                                continue
                    except IndexError:
                        #raise IndexError('while reading %s' % data)
                        print 'Error while reading: %s' % data #Just ignore that line
                        continue
                # Try to convert ids to int
                try:
                    row_id = int(row_id)
                except: pass
                try:
                    col_id = int(col_id)
                except: pass
                # Add tuple
                try:
                    self.add_tuple((float(value), row_id, col_id))
                except:
                    if VERBOSE:
                        sys.stdout.write('\nError while reading (%s, %s, %s). Skipping this tuple\n' % (value, row_id, col_id))
                    #raise ValueError('%s is not a float, while reading %s' % (value, data))
                i += 1
                if VERBOSE:
                    if i % 100000 == 0:
                        sys.stdout.write('.')
                    if i % 1000000 == 0:
                        sys.stdout.write('|')
                    if i % 10000000 == 0:
                        sys.stdout.write(' (%d M)\n' % int(i/1000000))
            if VERBOSE:
                sys.stdout.write('\n')

    def _load_pickle(self, path):
        """
        Loads data from a pickle file

        :param path: output filename
        :type param: string
        """
        self._data = pickle.load(codecs.open(path))

    def save(self, path, pickle=False):
        """
        Saves data in output file

        :param path: output filename
        :type param: string
        :param pickle: save in pickle format?
        :type pickle: Boolean
        """
        if VERBOSE:
            sys.stdout.write('Saving data to %s\n' % path)
        if pickle:
            self._save_pickle(path)
        else:
            out = codecs.open(path, 'w', 'utf8')
            for value, row_id, col_id in self._data:
                try:
                    value = unicode(value, 'utf8')
                except:
                    if not isinstance(value, unicode):
                        value = str(value)
                try:
                    row_id = unicode(row_id, 'utf8')
                except:
                    if not isinstance(row_id, unicode):
                        row_id = str(row_id)
                try:
                    col_id = unicode(col_id, 'utf8')
                except:
                    if not isinstance(col_id, unicode):
                        col_id = str(col_id)

                s = '\t'.join([value, row_id, col_id])
                out.write(s + '\n')
            out.close()

    def _save_pickle(self, path):
        """
        Saves data in output file, using pickle format

        :param path: output filename
        :type param: string
        """
        pickle.dump(self._data, open(path, "w"))
